src/lib.rs

```diff
@@ -287,69 +327,518 @@ impl<D: AsMut<[Digit]>> BigInt<D> {
+impl_digit_operation_reverse!(Mul, mul, u8);
+impl_digit_operation!(Mul, mul, u16);
+impl_digit_operation_reverse!(Mul, mul, u16);
+impl_digit_operation!(Mul, mul, u32);
+impl_digit_operation_reverse!(Mul, mul, u32);
+impl_u64_operation!(Mul, mul);
+impl_digit_operation_reverse!(Mul, mul, u64);
+impl_digit_operation_reverse!(Mul, mul, Digit);
+
+impl<D1, D2> Div<&BigInt<D2>> for &BigInt<D1>
+where
+    D1: AsRef<[Digit]>,
+    D2: AsRef<[Digit]>,
+{
     type Output = BigInt;
 
-    #[inline(always)]
-    fn add(self, rhs: &BigInt<D>) -> Self::Output {
-        rhs + self
+    /// Computes `floor(self / rhs)`
+    ///
+    /// ## Constant time
+    /// Division is constant time in size of its arguments. It means that the number of
+    /// instructions and memory access patterns remain the same for the same
+    /// `self.size()` and `rhs.size()` (provided that `rhs` does not start from 0).
+    ///
+    /// ## Zeroizing
+    /// This function might allocate some auxiliary buffer on the heap.
+    /// Before deallocating, it gets zeroed out.
+    ///
+    /// ## Complexity
+    /// The complexity is always `O(self.size() * rhs.size())`.
+    ///
+    /// ## Panics
+    /// The method panics if `rhs` equals zero.
+    fn div(self, rhs: &BigInt<D2>) -> Self::Output {
+        let n = self.size();
+        // we have to clone `self` in order to be able to use unchecked division,
+        // as it requires its first argument to be mutable
+        let mut a = self.reallocate();
+        let (a, b) = (a.as_mut(), rhs.as_ref());
+
+        // m is the length of b without leading zeroes
+        let m = ops::len_without_leading_zeroes(b);
+
+        if m == 0 {
+            panic!("Division by zero");
+        }
+
+        // if m > n, then definitely b > a and so floor(a/b) = 0.
+        if m > n {
+            a.zeroize();
+            return BigInt::zero().reallocate();
+        }
+
+        // Estimate the space we need to allocate on the heap
+        let output_size = n - m + 1;
+        let scratch_space_size = unsafe { ops::div_n_m_itch(n, m) };
+        debug_assert_eq!(scratch_space_size, ScratchSpace::division(n, m));
+
+        // Allocate space on the heap, split it into `output` that stands for addition result and
+        // `scratch_space` that is auxiliary buffer for gmp functions
+        let mut buffer = vec![Digit::zero(); output_size + scratch_space_size];
+        let (output, scratch_space) = buffer.split_at_mut(output_size);
+        let mut output = BigInt::from_digits(output);
+        let scratch_space = ScratchSpace::new(scratch_space);
+
+        let mut a_bigint = BigInt::from_digits(a);
+        unsafe {
+            // Safety:
+            // * We checked that n>=m and that n and m are non-zero (see match statement above)
+            // * We allocated sufficient memory for both output and scratch space
+            output.div_unchecked(&mut a_bigint, &BigInt::from_digits(&b[..m]), scratch_space);
```

Comment:
Size of `output` is `n - m + 1`. `div_unchecked` requires (as described in "safety" section) output to have size `n-m`.  I guess it's a typo in the "safety" section?

Reply:
Sloppy copy-pasting, thanks